R/content-tools.R

```diff
@@ -44,35 +42,62 @@ extract_tool_requests <- function(contents) {
   contents[is_tool_request]
 }
 
+tool_result_factory <- function(tool, arguments, id) {
+  function(result = NULL, error = NULL) {
+    if (S7_inherits(result, ContentToolResult)) {
+      result@id <- id
+      result@call_tool <- tool
+      result@call_args <- arguments
+      return(result)
+    }
+
+    ContentToolResult(
+      value = result,
+      error = error,
+      id = id,
+      call_tool = tool,
+      call_args = arguments
+    )
+  }
+}
+
 # Also need to handle edge cases: https://platform.openai.com/docs/guides/function-calling/edge-cases
-invoke_tool <- function(fun, arguments, id) {
-  if (is.null(fun)) {
-    return(ContentToolResult(id = id, error = "Unknown tool"))
+invoke_tool <- function(tool, arguments, id) {
```

Comment:
Given that we're now passing `tool` + `arguments` + `id` around in a few places, I wonder if it's worth making an object to represent this? It seems pretty close to the existing `ContentToolRequest`, so maybe that could gain a `function` property?  Then `ContentToolResult` could have a request property, which seems kind of nice?

Reply:
I like that too; it would definitely be cleaner to attach the `ContentToolRequest` to a `@request` property on the `ContentToolResult`.

Instead of adding `@function` to `ContentToolRequest`, I'd like to just add the tool def to `ContentToolRequest@tool`, and then we don't need the `@call_*` properties on the result object.